doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,1208 @@
+subclassing concrete DTypes directly:
+Logic such as ``can_cast`` does not cleanly inherit e.g. from a
+``Int64`` to a ``Datetime64`` even though the ``Datetime64`` could be seen
+as an integer with only a unit attached (and thus as a subclass).
+
+For the DType implementor this currently mainly means that DTypes are
+never subclasses of existing concrete DTypes.
+End-users should not notice or need to know about this distinction.
+
+The combination of (initially) rejecting subclassing of concrete DTypes
+while allowing it for abstract ones allows the transparent definition of
+a class hierarchy, while avoiding potential issues with subclassing and
+especially inheritance.
+
+Note that subclassing may be a possible mechanism to extend the datatypes
+in the future; not necessarily to implement new datatypes
+but rather to allow specialized implementations such as a GPU
+float64 subclassing a NumPy float64.
+
+As a technical implementation detail the DType class will require the C-side
+storage of methods and additional information.
+This requires the creation of a ``DTypeMeta`` class.
+Each ``DType`` class is thus an instance of ``DTypeMeta`` with a well defined
+and extensible interface.
+The end-user will not need to be aware of this.
+
+.. _hierarchy_figure:
+.. figure:: _static/dtype_hierarchy.svg
+    :figclass: align-center
+    
+
+Methods/Slots defined for each DType
+""""""""""""""""""""""""""""""""""""
+
+NEP 41 detailed that all logic should be defined through special methods
+on the DTypes.
+This section will list a specific set of such methods (in the form of 
+Python methods).
+The C-side equivalent slot signature will be summarized below after proposing
+the general C-API for defining new Datatypes.
+Note that while the slots are defined as special Python methods here, this is
+for convenience to the reader and *not* meant to imply the identical exposure
+as a Python API.
+This will need to be proposed in a separate, later, NEP.
+
+Some of the methods may be similar or even reuse existing Python slots.
+User-defined DType classes are discouraged from defining or using Python's
+special slots without feedback, in order to allow defining them later.
+For example ``dtype1 & dtype2`` (or maybe ``|``) could be a shorthand for
+``np.common_dtype``, and comparisons should be defined mainly through casting
+logic.
+
+
+Additional Information
+^^^^^^^^^^^^^^^^^^^^^^
+
+In addition to the more detailed methods below, the following general
+information is currently provided and will be defined on the class:
+
+* ``cls.parametric`` (see also `NEP 40 <NEP40>`_):
+
+  * This flag is mainly to simplify DType and casting creation and
+    allow for performance tweaks.
+  * DTypes which are not parametric must define a canonical dtype instance
+    which should be a singleton.
+  * Parametric dtypes require some additional methods (below).
+  * One can alternatively create a ``ParametricDType`` class for this purpose,
```

Comment:
Will look into the image (and other comments).

I think am OK with going there, the question is mainly how to expose it in the C-API I think, for the Python API, I am definitely +1, will add both as options for now. Internally, I think I can assume the flag exists, since I can always set it as part of the metaclass instantiation (similar to python using a flag to implement super-fast subclass checks for some basic types).